Ports/epsilon/patches/serenity.patch

```diff
@@ -0,0 +1,327 @@
++    if (strcmp(identifier, resources_addresses[i].identifier()) == 0) {
++      if (strcmp(jpgExtension, identifier + strlen(identifier) - strlen(jpgExtension)) == 0) {
++        format = AssetFormat::JPG;
++      } else {
++        assert(strcmp(pngExtension, identifier + strlen(identifier) - strlen(pngExtension)) == 0);
++        format = AssetFormat::PNG;
++      }
++      assetStart = resources_addresses[i].start();
++      assertSize = resources_addresses[i].end() - resources_addresses[i].start();
++      break;
++    }
++  }
++  assert(assetStart);
++
++  unsigned int width;
++  unsigned int height;
++  unsigned int bytesPerPixel;
++  unsigned char * bitmapData;
++  Uint32 pixelFormat;
++
++  switch (format) {
++    case AssetFormat::JPG:
++      if (!readJPG(assetStart, assertSize, &bitmapData, &width, &height, &bytesPerPixel, &pixelFormat)) {
++        return nullptr;
++      }
++      break;
++    default:
++      assert(format == AssetFormat::PNG);
++      if (!readPNG(assetStart, assertSize, &bitmapData, &width, &height, &bytesPerPixel, &pixelFormat)) {
++        return nullptr;
++      }
++  }
++
++  assert(bytesPerPixel == SDL_BYTESPERPIXEL(pixelFormat));
++
++  SDL_Texture * texture = SDL_CreateTexture(
++    renderer,
++    pixelFormat,
++    SDL_TEXTUREACCESS_STATIC,
++    width,
++    height
++  );
++
++  SDL_UpdateTexture(
++    texture,
++    nullptr,
++    bitmapData,
++    width * bytesPerPixel
++  );
++
++  SDL_SetTextureBlendMode(texture, SDL_BLENDMODE_BLEND);
++
++  delete[] bitmapData;
++
++  return texture;
++}
++
++class RGB888Pixel {
++public:
++  RGB888Pixel() {}
++  RGB888Pixel(KDColor c) :
++    m_red(c.red()),
++    m_green(c.green()),
++    m_blue(c.blue()) {
++  }
++private:
++  uint8_t m_red;
++  uint8_t m_green;
++  uint8_t m_blue;
++};
++static_assert(sizeof(RGB888Pixel) == 3, "RGB888Pixel shall be 3 bytes long");
++
++void saveImage(const KDColor * pixels, int width, int height, const char * path) {
++  FILE * file = fopen(path, "wb"); // Write in binary mode
++
++  png_structp png = png_create_write_struct(PNG_LIBPNG_VER_STRING, nullptr, nullptr, nullptr);
++  png_infop info = png_create_info_struct(png);
++  png_init_io(png, file);
++
++  png_set_IHDR(png, info,
++      width, height,
++      8, // Number of bits per channel
++      PNG_COLOR_TYPE_RGB,
++      PNG_INTERLACE_NONE,
++      PNG_COMPRESSION_TYPE_DEFAULT,
++      PNG_FILTER_TYPE_DEFAULT);
++
++  png_write_info(png, info);
++
++  RGB888Pixel * row = new RGB888Pixel[3*width];
++  for (int j=0;j<height;j++) {
++    for (int i=0; i<width; i++) {
++      row[i] = RGB888Pixel(pixels[i+width*j]);
++    }
++    png_write_row(png, reinterpret_cast<png_bytep>(row));
++  }
++  delete row;
++
++  png_write_end(png, NULL);
++
++  png_free_data(png, info, PNG_FREE_ALL, -1); // -1 = all items
++  png_destroy_write_struct(&png, nullptr);
++  fclose(file);
++}
++
++}
++}
++}
```

Comment:
Please split this into multiple patches (not necessarily one per file, but in a way that makes sense)